persistence/secrets.go

```diff
@@ -0,0 +1,296 @@
+package persistence
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/boltdb/bolt"
+	"github.com/golang/glog"
+	"github.com/open-horizon/anax/cutil"
+	"github.com/open-horizon/anax/exchangecommon"
+	"github.com/open-horizon/anax/semanticversion"
+	"time"
+)
+
+const SECRETS = "secrets"
+
+type PersistedServiceSecret struct {
+	SvcOrgid        string
+	SvcUrl          string
+	SvcArch         string
+	SvcVersionRange string
+	SvcSecretName   string
+	SvcSecretValue  string
+	AgreementIds    []string
+	TimeCreated     uint64
+	TimeLastUpdated uint64
+}
+
+type PersistedServiceSecrets struct {
+	MsDefId    string
+	MsDefVers  string // Version in the individual secrets is the range from the secret binding. This is the specific version associated with the msdef
+	MsDefUrl   string
+	MsDefOrg   string
+	SecretsMap map[string]PersistedServiceSecret
+}
+
+func PersistedSecretFromPolicySecret(inputSecretBindings []exchangecommon.SecretBinding, inputBoundSecrets []exchangecommon.BoundSecret, agId string) []PersistedServiceSecret {
+	boundSecFullMap := exchangecommon.BoundSecret{}
+	for _, boundSecMap := range inputBoundSecrets {
+		for key, val := range boundSecMap {
+			boundSecFullMap[key] = val
+		}
+	}
+
+	outputSecrets := []PersistedServiceSecret{}
+	for _, secBind := range inputSecretBindings {
+		for _, secArray := range secBind.Secrets {
+			for secName, _ := range secArray {
+				outputSecrets = append(outputSecrets, PersistedServiceSecret{SvcOrgid: secBind.ServiceOrgid, SvcUrl: secBind.ServiceUrl, SvcArch: secBind.ServiceArch, SvcVersionRange: secBind.ServiceVersionRange, SvcSecretName: secName, SvcSecretValue: boundSecFullMap[secName], AgreementIds: []string{agId}})
+			}
+		}
+	}
+
+	return outputSecrets
+}
+
+// Saves the given secret to the agent db
+func SaveSecret(db *bolt.DB, secretName string, msDefId string, msDefVers string, secretToSave PersistedServiceSecret) error {
+	secretToSaveAll := PersistedServiceSecrets{MsDefId: msDefId, MsDefOrg: secretToSave.SvcOrgid, MsDefUrl: secretToSave.SvcUrl, MsDefVers: msDefVers, SecretsMap: map[string]PersistedServiceSecret{}}
+	if allSecsForServiceInDB, err := FindAllSecretsForMS(db, msDefId); err != nil {
+		return fmt.Errorf("Failed to get all secrets for microservice %v. Error was: %v", msDefId, err)
+	} else if allSecsForServiceInDB != nil {
+		secretToSaveAll = *allSecsForServiceInDB
+	}
+
+	if secretToSave.TimeCreated == 0 {
+		secretToSave.TimeCreated = uint64(time.Now().Unix())
+	}
+
+	secretToSave.TimeLastUpdated = uint64(time.Now().Unix())
```

Comment:
Should this `secretToSave` be `PersistedServiceSecret` or `*PersistedServiceSecret`, since the `TimeCreated` and `TimeLastUpdated` field is updated in the function? 
I tested my secrets API and find after I saved 1 secret, the TimeCreate and TimeLastUpdated will not change even though I passed in different values. Maybe it is related?